packages/browser/src/profiling/utils.ts

```diff
@@ -365,3 +274,193 @@ export function convertJSSelfProfileToSampledFormat(input: JSSelfProfile): Threa
+      const file = stackFrame && stackFrame.filename;
+
+      if (stackFrame && file) {
+        acc[file] = debugIdMap[debugIdStackTrace] as string;
+        break;
+      }
+    }
+    return acc;
+  }, {});
+
+  const images: DebugImage[] = [];
+  for (const path of resource_paths) {
+    if (path && filenameDebugIdMap[path]) {
+      images.push({
+        type: 'sourcemap',
+        code_file: path,
+        debug_id: filenameDebugIdMap[path] as string,
+      });
+    }
+  }
+
+  return images;
+}
+
+/**
+ * Checks the given sample rate to make sure it is valid type and value (a boolean, or a number between 0 and 1).
+ */
+export function isValidSampleRate(rate: unknown): boolean {
+  // we need to check NaN explicitly because it's of type 'number' and therefore wouldn't get caught by this typecheck
+  if ((typeof rate !== 'number' && typeof rate !== 'boolean') || (typeof rate === 'number' && isNaN(rate))) {
+    __DEBUG_BUILD__ &&
+      logger.warn(
+        `[Profiling] Invalid sample rate. Sample rate must be a boolean or a number between 0 and 1. Got ${JSON.stringify(
+          rate,
+        )} of type ${JSON.stringify(typeof rate)}.`,
+      );
+    return false;
+  }
+
+  // Boolean sample rates are always valid
+  if (rate === true || rate === false) {
+    return true;
+  }
+
+  // in case sampleRate is a boolean, it will get automatically cast to 1 if it's true and 0 if it's false
+  if (rate < 0 || rate > 1) {
+    __DEBUG_BUILD__ &&
+      logger.warn(`[Profiling] Invalid sample rate. Sample rate must be between 0 and 1. Got ${rate}.`);
+    return false;
+  }
+  return true;
+}
+
+function isValidProfile(profile: ProcessedJSSelfProfile): profile is ProcessedJSSelfProfile & { profile_id: string } {
+  if (profile.samples.length <= 1) {
+    if (__DEBUG_BUILD__) {
+      // Log a warning if the profile has less than 2 samples so users can know why
+      // they are not seeing any profiling data and we cant avoid the back and forth
+      // of asking them to provide us with a dump of the profile data.
+      logger.log('[Profiling] Discarding profile because it contains less than 2 samples');
+    }
+    return false;
+  }
+
+  if (!profile.profile_id) {
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Creates a profiling envelope item, if the profile does not pass validation, returns null.
+ * @param event
+ * @returns {Profile | null}
+ */
+export function createProfilingEvent(profile: ProcessedJSSelfProfile, event: ProfiledEvent): Profile | null {
+  if (!isValidProfile(profile)) {
+    return null;
+  }
+
+  return createProfilePayload(event, profile);
+}
+
+
+export const PROFILE_MAP: Map<string, ProcessedJSSelfProfile> = new Map();
+/**
+ *
+ */
+export function addProfileToMap(profile: ProcessedJSSelfProfile): void{
+  PROFILE_MAP.set(profile.profile_id, profile);
+
+  if(PROFILE_MAP.size > 30){
+    const last: string = [...PROFILE_MAP.keys()].pop() as string;
+    PROFILE_MAP.delete(last);
+  }
```

Comment:
`Map` iteration is in insertion order (not sure if this applies to `.keys()` so wouldn't this pop off the profile you just inserted? If you want to stop adding profiles, then I think you want something like
```javascript
if (PROFILE_MAP.size > 30) {
  return;
}
PROFILE_MAP.set(profile.profile.id, profile);
```

On the other hand, it might make more sense to pop off the first key as that profile has been there the longest perhaps we should keep the newest profiles instead.

Reply:
No, I'll drop the first one in this case as it's been there the longest, that makes more sense